### PR TITLE
[DS-588] id field is based on business identification

### PR DESCRIPTION
**Pull request by cvriel**
On request of the user the value for the ID field is based on a business
resource identication. Before that a ID value was generated based on order of load.

The user needs a more steady and reliable key for comparision of different api calls
to detect changes.

### CODE REVIEW

**Comment by gkoller** [comment]
Like what you're doing here (`import_sport.py`). Lot's of type annotations, lots of docstrings. Nice!

File: src/dags/importscripts/import_sport.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import json
+import logging
+from typing import List, Tuple, TypedDict, Generator, Type
+
+
+# define logger for output to console
+logger = logging.getLogger(__name__)
+
+
+def get_dialect(file: str) -> Type:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:12]: Some type annotations can be a bit more specific. You're returning a `Dialect` type here, hence `-> Type[Dialect]`

File: src/dags/importscripts/import_sport.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    Returns:
+        A subclass dialect of Sniffer
+
+    """
+    with open(file) as data:
+        return csv.Sniffer().sniff(data.readline())
+
+
+def get_header(file: str) -> List:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:26]: `-> List[str]`

File: src/dags/importscripts/import_sport.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        a list of columns in string
+
+    """
+    dialect = get_dialect(file)
+    with open(file, "r") as src:
+        data = csv.reader(src, dialect=dialect)
+        return next(data)
+
+
+def make_hash(composite_values: List, digest_size: int = None) -> int:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:42]: `composite_values: List[str]`

And either: 
* `digest_size: int = 3` and remove the inline if statement in de function body (preferred) 
* or `digest_size: Optional[int] = None` and keep the inline if statement. (not recommend as you would need an explicit test for `is not None` to satisfy MyPy before passing `digest_size` to `blake2s`

File: src/dags/importscripts/import_sport.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    """The blake2s algorithm is used to generate a single hased value for source
+    composite values that uniquely identify a row.
+    In case the source itself doesn't provide a single solid identification as key.
+
+    Args:
+        composite_values: a concatenated string of record values based on
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:48]: The concatenation happens in the function body, hence the parameter is not yet concatenated.

File: src/dags/importscripts/import_sport.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                        which columns unqiuely identifiy a record
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:49]: uniquely

File: src/dags/importscripts/import_sport.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        digest_size: size to set the max bytes to use for the hash
+
+    Returns:
+        a hased value of the composite values in int
+
+    Note:
+        The default digest size is for now set to 3 bytes, which is equivant
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:56]: equivalent

File: src/dags/importscripts/import_sport.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        to ~ 10 karakters long.
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:57]: Actually a 3 byte `digest_size`  creates a 6 character long hex digest. Reason being that each hex digit encodes 4 bits (16 values). Each byte has 2 x 4 bits. Hence 3 bytes creates a 2 x 6 hex digit digest.

File: src/dags/importscripts/import_sport.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+        Because the id column is preferably of type int, the hased value is converted
+        from hex to int.
+    """
+    digest_size = 3 if not digest_size else digest_size
+    return int(
+        blake2s("|".join(composite_values).encode(), digest_size=digest_size).hexdigest(), base=16
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:63]: You are converting a digest in bytes to hex and then from hex to int. This can be done in one step:

```python
int.from_bytes(blake2s(...).digest())
```

Comment (gkoller) [reply src/dags/importscripts/import_sport.py:63]: `digest()` returns the internal byte representation (no conversion). `hexdigest()` does do a conversion.

File: src/dags/importscripts/import_sport.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    )
+
+
+def read_data(file: str, composite_key: Tuple) -> Generator:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:67]: `composite_key: Tuple[str]`

A `Generator` is [specified](https://docs.python.org/3.8/library/typing.html?highlight=generator#typing.Generator) as follows: `Generator[YieldType, SendType, ReturnType]`. If you are only yielding values you are better of using `Iterator`. 

First move the `DataObject` TypedDict definition out of the function body and specify `Iterator[DataObject]` as the return type of the function. 

Next use [DictReader](https://docs.python.org/3.8/library/csv.html?highlight=dictreader#csv.DictReader) and simplify the for loop as follows:

```python
for line in data:
    yield cast(DataObject, line)
```
This means you can also do away with `header = get_header(file)`

File: src/dags/importscripts/import_sport.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    """Read the data from csv file and contain it into a data class
+
+    Args:
+        file: the path/to/file.ext that will be read
+        composite_key: Tuple of column name(s) that unqiuely identifiy a record
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:72]: uniquely ;-)

File: src/dags/importscripts/import_sport.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        from hex to int.
+    """
+    digest_size = 3 if not digest_size else digest_size
+    return int(
+        blake2s("|".join(composite_values).encode(), digest_size=digest_size).hexdigest(), base=16
+    )
+
+
+def read_data(file: str, composite_key: Tuple) -> Generator:
+    """Read the data from csv file and contain it into a data class
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:68]: I'm sorry for nitpicking, but either you should use a real data class (as suggested by your docstring), or if you want to stick to the `TypedDict`, leave _and contain it into a data class_ out of the docstring. What the function returns is already specified in the return type hence you can be brief in your docstring.

File: src/dags/importscripts/import_sport.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+    Args:
+        file: the path/to/file.ext that will be read
+        composite_key: Tuple of column name(s) that unqiuely identifiy a record
+
+    Returns:
+        a generator containing the rows of the file plus it's header and the defined composite key
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:75]: It does not return a Generator. It yields row values. Replace `Returns` with `Yields` Eg.

```python
Yields:
   The next row from the data file.
```
See also: https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html (though no need to specify the type as they do there as we are using type annotations for that)

File: src/dags/importscripts/import_sport.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    header = get_header(file)
+    # setup the generic TypeDict data mold
+    DataObject = TypedDict("DataObject", {col: str for col in header})
+    with open(file, "r") as src:
+        data = csv.reader(src, dialect=dialect)
+        for line in data:
+            # construct lines into dictionary format, to proces into the generic data mold
+            row_dict = {i[0]: i[1] for i in zip(header, line)}
+            row_object = dict(object_=DataObject(row_dict), composite_key=composite_key)
+            yield row_object
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:88]: The `csv` mode has the [DictReader](https://docs.python.org/3.8/library/csv.html?highlight=dictreader#csv.DictReader). It does what you are trying to do here. By using `DictReader` you can simplify the for loop as follows:

```python
for line in data:
    yield cast(DataObject, line)
```
The `cast` is necessary as `DictReader` returns untyped dicts.

This means you can also do away with `header = get_header(file)`

File: src/dags/importscripts/import_sport.py (hunk 9 above)
Comment (gkoller) [reply src/dags/importscripts/import_sport.py:67]: Oeps, you can't move the `DataObject` definition out of the function body as its definition is dynamically created. I wouldn't do that as dynamically created data types don't do much for static analysis (or for you IDE of for other programmers.)

Just use `Iterator[Dict[str, Any]]`

File: src/dags/importscripts/import_sport.py (hunk 13 above)
Comment (gkoller) [reply src/dags/importscripts/import_sport.py:88]: See also https://github.com/Amsterdam/dataservices-airflow/pull/196/files#r580020020

`DataObject` might nog be necessary. If you use `Iterator[Dict[str, Any]` as your function return type. You might need still need a cast here, but now it becomes:
```python
yield cast(Dict[str, Any], line)
```

File: src/dags/importscripts/import_sport.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+
+def generate_unique_id(file: str, composite_key: Tuple, digest_size: int = None) -> Generator:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:91]: * `composite_key: Tuple[str]`
* `digest_size: Optional[int] = None`
* `-> Iterator[Dict[str, Any]]`

File: src/dags/importscripts/import_sport.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    Returns:
+        a generator containing the rows plus the unique ID value for the added ID column
+
+    """
+    for row in read_data(file, composite_key):
+        composite_values = []
+        for column in row["composite_key"]:
+            composite_values.append(row["object_"].get(column, None))
+        try:
+            row["object_"]["id"] = make_hash(composite_values, digest_size=digest_size)
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:108]: Because `make_hash`'s `digest_size` is no longer `Optional` (if you've followed my suggestions) this now becomes:

```python
options = {"digest_size": digest_size} if digest_size else {}
row["object_"]["id"] = make_hash(composite_values, **options)
```

File: src/dags/importscripts/import_sport.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+    rows = list()
+    for row in data:
+        rows.append(row.values())
+    with open(file, "w") as f:
+        write = csv.writer(f, dialect=csv.unix_dialect)
+        write.writerow(header)
+        write.writerows(rows)
+
+
+def add_unique_id_to_geojson(file: str, composite_key: Tuple, digest_size: int = None) -> None:
```
Comment (gkoller) [inline src/dags/importscripts/import_sport.py:138]: `composite_key: Tuple[str]`

You can remove `digest_size` as you are not using it.

File: src/dags/importscripts/import_sport.py (hunk 4 above)
Comment (gkoller) [reply src/dags/importscripts/import_sport.py:48]: On second thought I'm not sure my remark is valid. After all you are concatenating things on line 160 in you call to `make_hash`

File: src/dags/importscripts/import_sport.py (hunk 1 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast

File: src/dags/importscripts/import_sport.py (hunk 2 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast

File: src/dags/importscripts/import_sport.py (hunk 3 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast, gekozen voor digest_size: int = 3

File: src/dags/importscripts/import_sport.py (hunk 4 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Dit is een dubbele concatenate, heb deze over het hoofd gezien. Ik laat die in de hash functie staan.

File: src/dags/importscripts/import_sport.py (hunk 5 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Oeps, typo

File: src/dags/importscripts/import_sport.py (hunk 6 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Oeps, typo

File: src/dags/importscripts/import_sport.py (hunk 7 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Je hebt gelijk, ik had eerst een digist_size van 5. Wat is gelijk aan 10 karakters. De tekst is niet aangepast. Van een 4 bits codering naar bytes is i.d.d. 3 / 0,5 = 6 karakters

File: src/dags/importscripts/import_sport.py (hunk 8 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast

File: src/dags/importscripts/import_sport.py (hunk 10 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: At least the typo is consistent ;-)

File: src/dags/importscripts/import_sport.py (hunk 12 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Eens, aangepast

File: src/dags/importscripts/import_sport.py (hunk 14 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast

File: src/dags/importscripts/import_sport.py (hunk 15 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Aangepast

File: src/dags/importscripts/import_sport.py (hunk 16 above)
Comment (cvriel) [reply src/dags/importscripts/import_sport.py]: Oeps, foutje
